hadrons/initial_recombination.py

```diff
@@ -0,0 +1,204 @@
+    negative_array_temp = np.zeros((no_x, no_x,no_z_with_buffer))
+    no_recombined_charge_carriers = 0.0
+    no_initialised_charge_carriers = 0.0
+
+    # for the colorbars
+    MAXVAL = Gaussian_factor
+
+    dt = 1.
+    von_neumann_expression = False
+    Efield = voltage_V/electrode_gap_cm
+
+    # find a time step dt which fulfils the von Neumann criterion, i.e. ensures the numericl error does not increase but
+    # decreases and eventually damps out
+    while not von_neumann_expression:
+        dt /= 1.01
+        # as defined in the Deghan (2004) paper
+        sx = ion_diff*dt/(unit_length_cm**2)
+        sy, sz = sx, sx
+
+        cx = ion_mobility*Efield*dt/unit_length_cm*sin(IC_angle_rad)
+        cy = 0
+        cz = ion_mobility*Efield*dt/unit_length_cm*cos(IC_angle_rad)
+
+        # check von Neumann's criterion
+        von_neumann_expression = (2*(sx + sy + sz) + cx**2 + cy**2 + cz**2 <= 1 and cx**2*cy**2*cz**2 <= 8*sx*sy*sz)
+
+    # calculate the number of step required to drag the two charge carrier distributions apart
+    separation_time_steps = int(electrode_gap_cm/(2.*ion_mobility*Efield*dt))
+    computation_time_steps = separation_time_steps*2
+
+    if debug:
+        print("Electric field = %s V/cm" % Efield)
+        print("Separation time = %3.2E s" % (separation_time_steps*dt))
+        print("Time steps      = %3.2E" % computation_time_steps)
+        print("Time step dt    = %3.2E s" % dt)
+        print("Number of voxels = %3.2E" % (no_x**2 * no_z_with_buffer))
+        print("Number of pixels = %d (x = y directions)" % no_x)
+        print("Number of pixels = %d (z direction)" % no_z)
+
+    # define the radial dose model to be used
+    if RDD_model == "Gauss":
+        def RDD_function(r_cm):
+            return Gaussian_factor * exp(-r_cm ** 2 / track_radius_cm ** 2)
+
+    elif RDD_model == "Geiss":
+        def RDD_function(r_cm):
+            return Geiss_RRD_cm(r_cm, c, a0_cm, r_max_cm)
+    else:
+        print("RDD model {} undefined.".format(RDD_model))
+        return 0
+
+    # initialise the Gaussian distribution in the array
+    start_time = time.time()
+    for i in range(no_x):
+        for j in range(no_x):
+            distance_from_center_cm = sqrt((i - mid_xy_array) ** 2 + (j - mid_xy_array) ** 2) * unit_length_cm
+            ion_density = RDD_function(distance_from_center_cm)
+            for k in range(no_z_electrode, no_z + no_z_electrode):
+                if debug: print(f'\rGaussian distribution loop: k - {k+1: >3}/{no_z + no_z_electrode}, i - {i+1: >3}/{no_x}, j - {j+1: >3}/{no_x} Time: {time.time()-start_time: .2f}', end='')
+                # ion_density = Gaussian_factor * exp(-distance_from_center ** 2 / track_radius_cm ** 2)
+                # ion_density = Geiss_RRD_cm(distance_from_center, c, a0_cm, r_max_cm)
+                no_initialised_charge_carriers += ion_density
+                positive_array[i, j, k] += ion_density
```

Comment:
what about NumPy implementation and moving to one level up ?
I am not sure what would work better with numba - pure python or numpy constructd
```suggestion
                positive_array[i, j, :] += ion_density
```